cincinnati/src/plugins/internal/edge_add_remove.rs

```diff
@@ -128,18 +132,30 @@ impl EdgeAddRemovePlugin {
     ///
     /// The labels are assumed to have the syntax `<prefix>.(previous|next).add=(<Version>,)*<Version>`
     fn add_edges(&self, graph: &mut cincinnati::Graph) -> Fallible<()> {
+        macro_rules! handle_add_edge {
+            ($direction:expr, $from:ident, $to:ident, $from_string:ident, $to_string:ident) => {
+                if let Err(e) = graph.add_edge(&$from, &$to) {
+                    if let Some(eae) = e.downcast_ref::<crate::errors::EdgeAlreadyExists>() {
+                        warn!("{}", eae);
+                        continue;
+                    };
+                    bail!(e);
+                };
+            };
+        }
+
         graph
             .find_by_metadata_key(&format!("{}.{}", self.key_prefix, "previous.add"))
             .into_iter()
-            .try_for_each(|(to, to_version, from_csv)| -> Fallible<()> {
-                for from_version in from_csv.split(',').map(str::trim) {
-                    if let Some(from) = graph.find_by_version(&from_version) {
-                        info!("[{}]: adding previous {}", &from_version, &to_version);
-                        graph.add_edge(&from, &to)?
+            .try_for_each(|(to, to_string, from_csv)| -> Fallible<()> {
```

Comment:
It looks like this `try_for_each` was drifting rightward and you had to introduce a macro. Did you maybe consider splitting this whole closure to its own function?

Reply:
I tried to use the macro to reduce the duplication between the closures for the previous and next. A function wouldn't help with that use-case, and I wasn't sure how much effort I want to put into making one macro for both cases.

Reply:
I guess we can leave this as-is for now then (I think that my minor itches here are the two macro-in-function).

If/when we need to have the same permissive add/remove logic in some other place, we can reconsider how to better share this logic.